cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +48,1501 @@ using parquet::detail::ColumnChunkDesc;
+                   "Second operand of binary operation with column reference must be a literal");
+      v->accept(*this);
+
+      if (op == ast_operator::EQUAL or op == ast::ast_operator::NOT_EQUAL) {
+        // Search the literal in this input column's equality literals list and add to
+        // the offset.
+        auto const col_idx            = v->get_column_index();
+        auto const& equality_literals = _literals[col_idx];
+        auto col_literal_offset       = _col_literals_offsets[col_idx];
+        auto const literal_iter       = std::find(equality_literals.cbegin(),
+                                            equality_literals.cend(),
+                                            dynamic_cast<ast::literal const*>(&operands[1].get()));
+        CUDF_EXPECTS(literal_iter != equality_literals.end(), "Could not find the literal ptr");
+        col_literal_offset += std::distance(equality_literals.cbegin(), literal_iter);
+
+        auto const& value = _dictionary_expr.push(ast::column_reference{col_literal_offset});
+
+        if (op == ast_operator::NOT_EQUAL) {
+          // For NOT_EQUAL operator, simply evaluate boolean is_false(value) expression as
+          // NOT(value). The value indicates if the row group should be pruned (if the literal is
+          // present in the hash set and it's the only value in the hash set)
+          _dictionary_expr.push(ast::operation{ast_operator::NOT, value});
+        } else {
+          // For EQUAL operator, evaluate boolean is_true(value) expression as IDENTITY(value)
+          // The value indicates if the row group should be kept (if the literal is present in the
+          // hash set)
+          _dictionary_expr.push(ast::operation{ast_operator::IDENTITY, value});
+        }
+      }
+      // For all other expressions, push the `always true` expression
+      else {
+        _dictionary_expr.push(ast::operation{ast_operator::IDENTITY, _always_true});
+      }
+    } else {
+      auto new_operands = visit_operands(operands);
+      if (cudf::ast::detail::ast_operator_arity(op) == 2) {
+        _dictionary_expr.push(ast::operation{op, new_operands.front(), new_operands.back()});
+      } else if (cudf::ast::detail::ast_operator_arity(op) == 1) {
+        _dictionary_expr.push(ast::operation{op, new_operands.front()});
+      }
+    }
+    return _dictionary_expr.back();
+  }
+
+  /**
+   * @brief Returns the AST to apply on dictionary membership.
+   *
+   * @return AST operation expression
+   */
+  [[nodiscard]] std::reference_wrapper<ast::expression const> get_dictionary_expr() const
+  {
+    return _dictionary_expr.back();
+  }
+
+ private:
+  std::vector<cudf::size_type> _col_literals_offsets;
+  cudf::host_span<std::vector<ast::literal*> const> _literals;
+  ast::tree _dictionary_expr;
+  cudf::numeric_scalar<bool> _always_true_scalar{true};
+  ast::literal const _always_true{_always_true_scalar};
```

Comment:
nit: is `_always_true_scalar` used anywhere else besides initializing the AST literal? If not, having a data member used only once to initialize another isn't worth the cost.
```suggestion
  ast::literal const _always_true{cudf::numeric_scalar<bool>{true}};
```

Reply:
Reverting this as `ast::literal const _always_true{}` takes in an lvalue (&scalar) so we can't iniitalize it like suggested